### ISSUE TITLE
Communication loss with upper devices in stack

### ISSUE THREAD

**Issue by aaronbeekay** [BuckeyeCurrent/Battery-Interface-Module#7]
On some occasions, some or all of the BQ ICs north of the first IC will stop responding to SPI communications. When the issue occurs, the devices clock back only `0x00`.

A representative test case, taken from a logic analyzer recording when the issue was occurring:
```
MOSI				MISO
7F 3C A5 57			00 00 00 00			// stack reset
00 00 01 00 00		00 00 00 61 35		// read dev. 1 status register - OK
01 3B 01 02			00 00 00 00			// assign dev. 1 address #1
02 3B 01 00 00		00 00 00 81 BA		// check dev. 1 address assigned - OK
00 00 01 00 00		00 00 00 00 00		// read dev. 2 status register - no reply
```

Any reads to devices north of device 1 will have the same result.

**Comment by aaronbeekay** [comment]
@lordnathan0 and I looked into this issue more today. Our findings:

* When we saw the issue today, it was related to the stack being put into SLEEP mode. When the stack is power cycled and initialized (addressed from an unaddressed state), it replies normally until the devices are put into SLEEP mode. After that point, communications with upper devices fail, even when the IO_CONTROL[SLEEP] bit is cleared.

Here is an example of a stack communications failure just after wake-from-sleep: [Wake-from-sleep communications error on BQ76PL536A stack](https://gist.github.com/aaronbeekay/1504642866915796bcae).

* Power cycling the stack revives it until it is put back to sleep again.
* Power cycling the host processor does not seem to do anything.

**Comment by aaronbeekay** [comment]
@lordnathan0 - when I was looking through the communications we recorded today, I noticed that we don't actually obey the minimum wake delay time for at least two devices:

![bq - wake from sleep timing](https://cloud.githubusercontent.com/assets/7378316/6746386/a637b178-ce9d-11e4-9ee3-df28bb6df03e.png)

I'm not sure if this is the root cause of the wake-from-sleep issue, since device 2 technically satisfies TI's "wait 1 ms" criterion. But it might be worth adding a delay for – the behavior isn't specified if you don't wait for their magic 1ms.

**Comment by aaronbeekay** [comment]
Also, in that Gist, there's another software bug: the algorithm correctly reads `ALERT_STATUS` in response to the `DEVICE_STATUS[ALERT]` bit being set, but then it doesn't react to the `ALERT_STATUS[PARITY]` flag being set:

```
MOSI				MISO
02 00 01 00 00		00 00 00 A1 57		// Read DEVICE_STATUS on dev. 1:
										//   DEVICE_STATUS[ALERT] = 1
										//	 DEVICE_STATUS[DRDY] = 1
						
02 20 01 00 00		00 00 00 40 BD		// Read ALERT_STATUS on dev. 1:
										//   ALERT_STATUS[PARITY] = 1 (invalid parity)
```

I can make an issue for this... @lordnathan0 where is the repo where BIM software is being developed?

**Comment by aaronbeekay** [comment]
It now appears that the issue occurs only after some delay after `IO_CONTROL[SLEEP]` is set. Reads from the upper devices succeed 10 ms after SLEEP entry, but fail 500 ms after.

We have posted a [thread on the TI E2E forums](https://e2e.ti.com/support/power_management/battery_management/f/1002/t/412204) seeking support.

**Comment by aaronbeekay** [comment]
Per Roger's comments on that E2E forum thread, it seems likely that the issue is that `HSEL` on the upper devices is pulled up to n-REG50. Since REG50 is disabled in SLEEP and HSEL has a pulldown resistor, the devices do not accept communications on their south interfaces after SLEEP mode entry.

Roger suggests pulling HSEL to n-LDOD.

**Comment by aaronbeekay** [comment]
Note: this issue has probably been in existence since BMM r01, several years ago. The likely source of the error is the following note in the BQ76PL536 datasheet (SLUSA08A) (emphasis mine):

>Designer Note: Because the LDODx inputs are pulled briefly to ~7 V during programming, the LDODx pins should not be used as sources for pullups to 5-V digital pins, such as HSEL and SPI(bus)_H connected pins. **Use LDOA or VREG50 instead**, unless all programming is completed prior to mounting on the application PCB, in which case LDODx is a good choice.

In later versions of the datasheet, this note was modified slightly, so that the text "Use LDOA or VREG50 instead" reads "Use VREG50 instead".

Not much to be done about that. Just a little lesson in blindly trusting datasheet recommendations.

**Comment by aaronbeekay** [comment]
This is fixed with a trace cut in v2.0.0. Need to fix board file. Target v2.0.2.
